certbot/cert_manager.py

```diff
@@ -20,3 +30,83 @@ def update_live_symlinks(config):
             update_symlinks=True)
+
+def _report_lines(msgs):
+    """Format a results report for a category of single-line renewal outcomes"""
+    certinfo = [str(msg) for msg in msgs]
+    return "  " + "\n  ".join(certinfo)
+
+def _report_human_readable(parsed_certs):
+    """Format a results report for a parsed cert"""
+    certinfo = []
+    for cert in parsed_certs:
+        now = pytz.UTC.fromutc(datetime.datetime.utcnow())
+        if cert.target_expiry <= now:
+            expiration_text = "EXPIRED"
+        else:
+            diff = cert.target_expiry - now
+            if diff.days == 1:
+                expiration_text = "1 day"
+            elif diff.days < 1:
+                expiration_text = "under 1 day"
+            else:
+                expiration_text = "{0} days".format(diff.days)
+        valid_string = "{0} ({1})".format(cert.target_expiry, expiration_text)
+        out = "  Lineage: {0}\n    Domains: {1}\n    Valid Until: {2}".format(
+            cert.lineagename, " ".join(cert.names()), valid_string)
+        certinfo.append(out)
+    return "\n".join(certinfo)
+
+def _describe_certs(config, parsed_certs, parse_failures):
+    """Print information about the certs we know about"""
+    out = []
+
+    notify = out.append
+
+    if not parsed_certs and not parse_failures:
+        notify("No certs found.")
+    else:
+        if parsed_certs:
+            notify("Found the following certs:")
+            notify(_report_human_readable(parsed_certs))
+        if parse_failures:
+            notify("\nThe following renewal configuration files "
+               "were invalid:")
+            notify(_report_lines(parse_failures))
+
+    if config.quiet:
+        return
+    disp = zope.component.getUtility(interfaces.IDisplay)
+    disp.notification("\n".join(out), pause=False)
+
+def list_certs(config):
+    """Display information about the certificates that Certbot knows about
+
+    :param config: Configuration.
+    :type config: :class:`certbot.interfaces.IConfig`
+    """
+    renewer_config = configuration.RenewerConfiguration(config)
+    parsed_certs = []
+    parse_failures = []
+    for renewal_file in renewal.renewal_conf_files(renewer_config):
+        lineage_config = copy.deepcopy(config)
+
+        # Note that this modifies config (to add back the configuration
+        # elements from within the renewal configuration file).
+        try:
+            renewal_candidate = storage.RenewableCert(renewal_file,
+                configuration.RenewerConfiguration(lineage_config))
+        except Exception as e:  # pylint: disable=broad-except
+            logger.warning("Renewal configuration file %s produced an "
+                           "unexpected error: %s. Skipping.", renewal_file, e)
+            logger.debug("Traceback was:\n%s", traceback.format_exc())
+            parse_failures.append(renewal_file)
+            continue
+
+        if renewal_candidate is None:
```

Comment:
That's fine. Appending to the list we made at the top of function really should never fail. If it does though, we'll incorrectly report the problem as the user's renewal conf file rather than a bug in our code or Python.

Like I said, this should never happen though, so I don't really have a problem if you take this approach.